### ISSUE TITLE
Improve generated unknown data type javadoc when forcedTypes match return type, but there are still unknown parameter types

### ISSUE THREAD

**Issue by perlun** [jOOQ/jOOQ#14255]
### Expected behavior

When overriding the return type for a function to something else than `Object`, no `@Deprecated` annotation should be emitted even though `withDeprecationOnUnknownTypes(true)` is used.

### Actual behavior

The method is annotated as `@Deprecated` in the generated `Routines` class.

### Steps to reproduce the problem

The use case is trying to use the TimescaleDB [`locf`](https://docs.timescale.com/api/latest/hyperfunctions/gapfilling-interpolation/locf/) function with jOOQ. The method returns `anyelement` which is what makes it a bit less straightforward.

Before trying to workaround this, the entry in `Routines.java` looks like this:

```java
    /**
     * @deprecated Unknown data type. If this is a qualified, user-defined type,
     * it may have been excluded from code generation. If this is a built-in
     * type, you can define an explicit {@link org.jooq.Binding} to specify how
     * this type should be handled. Deprecation can be turned off using
     * {@literal <deprecationOnUnknownTypes/>} in your code generator
     * configuration.
     */
    @Deprecated
    public static Object locf(
          Configuration configuration
        , Object value
        , Object prev
        , Boolean treatNullAsMissing
    ) {
        Locf f = new Locf();
        f.setValue(value);
        f.setPrev(prev);
        f.setTreatNullAsMissing(treatNullAsMissing);

        f.execute(configuration);
        return f.getReturnValue();
    }
```

After adding a`ForcedType` like this (can provide the full generator invocation on demand):

```java
.withForcedTypes(
        // `locf` needs special treatment, as it is defined with
        // a return type as `anyelement`.
        new ForcedType()
                .withName( "NUMERIC" )
                .withIncludeExpression( "locf.RETURN_VALUE" )
)
```

...we now get something like this. The data type override is correctly applied, but the deprecation still remains:

```java
    /**
     * @deprecated Unknown data type. If this is a qualified, user-defined type,
     * it may have been excluded from code generation. If this is a built-in
     * type, you can define an explicit {@link org.jooq.Binding} to specify how
     * this type should be handled. Deprecation can be turned off using
     * {@literal <deprecationOnUnknownTypes/>} in your code generator
     * configuration.
     */
    @Deprecated
    public static BigDecimal locf(
          Configuration configuration
        , Object value
        , Object prev
        , Boolean treatNullAsMissing
    ) {
        Locf f = new Locf();
        f.setValue(value);
        f.setPrev(prev);
        f.setTreatNullAsMissing(treatNullAsMissing);

        f.execute(configuration);
        return f.getReturnValue();
    }
```

Note that this only pertains to the `Routines.java` file; the actual `routines/Locf.java` looks much better (no `@Deprecated`):

```diff
diff --git centre/audit-db-schema/src/generated/java/fi/hibox/centre/audit/db/domain/routines/Locf.java centre/audit-db-schema/src/generated/java/fi/hibox/centre/audit/db/domain/routines/Locf.java
index 658d480712..e2abd26ccf 100644
--- centre/audit-db-schema/src/generated/java/fi/hibox/centre/audit/db/domain/routines/Locf.java
+++ centre/audit-db-schema/src/generated/java/fi/hibox/centre/audit/db/domain/routines/Locf.java
@@ -6,6 +6,8 @@ package fi.hibox.centre.audit.db.domain.routines;
 
 import fi.hibox.centre.audit.db.domain.Public;
 
+import java.math.BigDecimal;
+
 import org.jooq.Field;
 import org.jooq.Parameter;
 import org.jooq.impl.AbstractRoutine;
@@ -14,30 +16,18 @@ import org.jooq.impl.Internal;
 import org.jooq.impl.SQLDataType;
 
 
-
 /**
- * @deprecated Unknown data type. If this is a qualified, user-defined type, it
- * may have been excluded from code generation. If this is a built-in type, you
- * can define an explicit {@link org.jooq.Binding} to specify how this type
- * should be handled. Deprecation can be turned off using {@literal
- * <deprecationOnUnknownTypes/>} in your code generator configuration.
+ * This class is generated by jOOQ.
  */
-@Deprecated
 @SuppressWarnings({ "all", "unchecked", "rawtypes" })
-public class Locf extends AbstractRoutine<Object> {
+public class Locf extends AbstractRoutine<BigDecimal> {
 
     private static final long serialVersionUID = 1L;
 
     /**
-     * @deprecated Unknown data type. If this is a qualified, user-defined type,
-     * it may have been excluded from code generation. If this is a built-in
-     * type, you can define an explicit {@link org.jooq.Binding} to specify how
-     * this type should be handled. Deprecation can be turned off using
-     * {@literal <deprecationOnUnknownTypes/>} in your code generator
-     * configuration.
+     * The parameter <code>public.locf.RETURN_VALUE</code>.
      */
-    @Deprecated
-    public static final Parameter<Object> RETURN_VALUE = Internal.createParameter("RETURN_VALUE", org.jooq.impl.DefaultDataType.getDefaultDataType("\"pg_catalog\".\"anyelement\""), false, false);
+    public static final Parameter<BigDecimal> RETURN_VALUE = Internal.createParameter("RETURN_VALUE", SQLDataType.NUMERIC, false, false);
 
     /**
      * @deprecated Unknown data type. If this is a qualified, user-defined type,
@@ -70,7 +60,7 @@ public class Locf extends AbstractRoutine<Object> {
      * Create a new routine call instance
      */
     public Locf() {
-        super("locf", Public.PUBLIC, org.jooq.impl.DefaultDataType.getDefaultDataType("\"pg_catalog\".\"anyelement\""));
+        super("locf", Public.PUBLIC, SQLDataType.NUMERIC);
 
         setReturnParameter(RETURN_VALUE);
         addInParameter(VALUE);
```

### jOOQ Version

jOOQ Professional Edition 3.17.0

### Database product and version

https://hub.docker.com/r/timescale/timescaledb, image tag `timescale/timescaledb:2.5.0-pg13-oss`

### Java Version

openjdk version "11.0.17" 2022-10-18

### OS Version

Debian GNU/Linux bookworm/sid

### JDBC driver name and version (include name if unofficial driver)

org.postgresql:postgresql:42.3.1

**Comment by lukaseder** [comment]
Thanks a lot for your report. I'll look into this soon

**Comment by lukaseder** [comment]
I see, I can reproduce this when using `locf.RETURN_VALUE`, but not with `locf.*`. I tested this on ordinary PG with this function:

```sql
CREATE FUNCTION f_any_forced(in_value IN anynonarray)
RETURNS anynonarray
AS $$
BEGIN
    RETURN in_value;
END;
$$ LANGUAGE plpgsql;
```

So, `f_any_forced.*` is a reasonable regex, matching both parameters and return value, but this obviously doesn't work in your case.

**Comment by lukaseder** [comment]
The problem also disappears when adding another `forcedType`, i.e. when at least one of the parameters also matches:

```xml
<forcedType>
    <name>BIGINT</name>
    <includeExpression>f_any_forced\.RETURN_VALUE</includeExpression>                                                
</forcedType>
<forcedType>
    <name>INTEGER</name>
    <includeExpression>f_any_forced\.IN_VALUE</includeExpression>                                                
</forcedType>
```

That could be helpful as a workaround for you?

**Comment by lukaseder** [comment]
A yes, duh. Your parameters are also of an unknown type, so the deprecation is valid. I guess the deprecation message could be a bit more clear in this particular case. Unlike in the routine class itself, the message can refer to both parameter types or return types in the `Routines` class

**Comment by perlun** [comment]
> A yes, duh. Your parameters are also of an unknown type, so the deprecation is valid. I guess the deprecation message could be a bit more clear in this particular case. Unlike in the routine class itself, the message can refer to both parameter types or return types in the `Routines` class

Ah... valid point in fact. :thinking: : So the way I override it is probably not entirely right, since I want it to be `NUMERIC` for _both_ the parameter and the return type in this case.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.6 (https://github.com/jOOQ/jOOQ/issues/14265), and 3.16.12 (https://github.com/jOOQ/jOOQ/issues/14266)
